---
title: "Project_4"
author: "Sarah Eiben, Michelle Baumgartner, Deni Hodzic"
date: '2022-06-14'
output: html_document
---

# Clustering of the signals and enrichment analysis

Our second approach to compare the different family members with eachother is to cluster the signals and do an enrichment analysis with the clusters. After plotting the signals in a combined heatmap, the signals will be clustered and analyzed in an enrichment analysis.


```{r}
suppressPackageStartupMessages({
  library(GenomicRanges)
  library(epiwraps)
  library(ggplot2)
  library(rGREAT)
  library(AnnotationHub)
  library(Rsubread)
  library(rtracklayer)
  library(Biostrings)
  library(Rfastp)
  library(ensembldb)
  library(bsseq)
  library(BiocParallel)
  library(edgeR)
  library(DMRcate)
})
```

## Plotting the signals

As regions we are going to be using the peaks from the different CREB family members that we downloaded in Project_2 and the tracks are the corresponding signal p-values.
We make a list of the signal p-values of the CREB family members.
Then we also make a list of the peak values and merge them together to create one long list of non-overlapping regions.
Finally we combine these in a matrix and create a heatmap.

```{r}
# create a list with all the peaks on chromosome 1
peak_list <- GRangesList("Peaks CREB1"=peaks_creb1_chr1, "Peaks CREB3"=peaks_creb3_chr1, "Peaks CREB3L1"=peaks_creb3l1_chr1, "Peaks CREB5"=peaks_creb5_chr1, compress=TRUE)
# merge the peaks together to create one list of non-overlapping regions
merged_peaks <- reduce(unlist(peak_list))
# make a list of the signal p-values
pv_list <- list(CREB1="signal_pv_creb/signal_pv_creb1.bw", CREB3="signal_pv_creb/signal_pv_creb3.bw", CREB3L1="signal_pv_creb/signal_pv_creb3l1.bw", CREB5="signal_pv_creb/Signal_pv_creb5.bw")
# we obtain the matrix of the signal around the regions:
m_creb1 <- signal2Matrix(pv_list, merged_peaks)
# maybe set an extend and a width?
plotEnrichedHeatmaps(m_creb1)
```
Each row represents the same region across the heatmaps.

Here you can see that there are strongest peaks in CREB3L1, followed by similar looking CREB1 and CREB3 and rather weak looking CREB5.

Analyze this more (not sure if I can speak of strengths of peaks here or what would be correct?).

## Clustering

Now we use clustering to set rows that are similar together and seperate them from the rest.
Let's start by trying with 5 clusters. Then we change the number of clusters until we reach a plateau of the variance explained by clusters.


```{r}
cl <- clusterSignalMatrices(m_creb1, k=10)
# with 5 clusters 73% of the variance is explained by clusters
# with 4 clusters 64 % of the variance is explained by clusters
# with 6 clusters 77%
# with 7 clusters 80%
# with 8 clusters 82%
# with 9 clusters 84%
# with 10 clusters 85%
# with 11 clusters 86%
# so let's go with 10 clusters and see what we can tell.
table(cl)
head(cl)
length(cl)
# The numbers at [1] tell you where your regions are: first region is in cluster 10 with 303 elements
merged_peaks$cluster <- cl
```

### Plotting the clusters

We can split the rows by clusters to visualize this.
The meaning of the cluster is dependent on the order of the matrix, which depends on the order of the rows in the regions.
It's really easy if one generates another regions list that is ordered  and one keeps the same cluster identities everything will be messed up!
We have to keep in mind that the orders of both of them need to be matched when something is changed.
Typically when clusters are exhausting most of the pattern in the data, the results are more stable. For reproducibility of workflows it is important to set the seed, we chose to set the seed at 123.

```{r}
set.seed(123)
plotEnrichedHeatmaps(m_creb1, row_split=cl)
```
Interpretation:

Cluster 1: Strongly represented in CREB1 and CREB3L1.
Cluster 2: Weak representation in all but CREB3.
Cluster 3: Medium representation in all but CREB5.
Cluster 4: Strong in CREB3 and one of the few that is also represented quite strongly in CREB5 (still pretty weak).
Cluster 5: Strong in CREB1, all the others weak or barely any signal.
Cluster 6: Strong in CREB1 and CREB3L1.
Cluster 7: Similar to Cluster 6, a little stronger in CREB3.
Cluster 8: Strong representation in CREB3 and weak in CREB1.
Cluster 9: Very strong and broad with few regions in CREB3.
Cluster 10: Similar to 6 and 7 although a bit stronger.

So overall CREB1 is strongly present 7/10 times, CREB3 5/10 times, CREB3L1 6/10 times and CREB5 has a medium strong to weak presence 2/10 times.

The curves at the top is the average signal across all the regions but with clusters we have multiple lines. But we cannot tell which one is which.
So we set colors.

#### Adding colors

```{r}
# create a vector of colors and have as names of the vector the names of our clusters (in this case just numbers).
mycolors <- c("1"="yellow", "2"="orange", "3"="red", "4"="pink", "5"="purple", "6"="blue", "7"="green", "8"="brown", "9"="gray", "10"="black")
plotEnrichedHeatmaps(m_creb1, row_split=cl, mean_color=mycolors, trim=c(0.95))
```

Now from the curves you can see that CREB1 stands out in cluster 7 and 10, CREB3 in cluster 2, 4 and 9, CREB3L1 in cluster 7 and 10 and CREB5 in cluster 2.

Generally said, CREB1 and CREB3L1 have rather narrow peaks and CREB3 tends to be more broad. CREB5 is not very present.

## Plotting just the averages

We tried to extract the average signals to get a more broad overview of the signals, we can also do this by cluster.
This gives us a data frame that has for each cluster for each sample or signal at each position the value of the mean signal.
We can use this to create a plot that tells us in each cluster what the average signals of each factor around the positions around the regions of interest are.

```{r}
d <- meltSignals(m_creb1, splitBy=cl)
ggplot(d, aes(position, mean, colour=sample)) + geom_line(size=1.2) + facet_wrap(~split)
```

These plots are exactly like the plots above the heat maps with the difference that instead of plotting the different lines for one signal together we are plotting the lines for one cluster but different signals together. You can plot whatever you want.

Clustering is dangerous because you might change around factors until you like the result.

From this plot we can generally tell once more that CREB5 is not very strongly involved in any of these peaks. All the other family members have some clusters that they are the strongest representative which means they might be strongly involved in whatever these processes represent.

From these curves and the ones above the heatmap, we decided to further analyse 3 of these 10 clusters - Cluster 2, Cluster 7 and Cluster 9.
Even though cluster 10 also has strong signals, we decided against further analysis because it looks similar to Cluster 7.

# Enrichment analysis

Now we want to figure out what the regions are. For example what are the regions in cluster 2, are they associated with or close to genes that have a different kind of function than the regions in cluster 7?
Typically you add more signals in order to find more patterns of correlation. Are they bound by other factors?

Find what's enriched in one cluster with respect to the others (first we use cluster 2):

```{r}
# we first split the regions by cluster:
split_regions <- split(merged_peaks, cl)
# Important: when a job is submitted to GREAT a set of regions, a GRanges object, needs to be submitted BUT there also needs to be some sort of background added to the picture. The elements in a list of regions are always compared to other elements. So we are comparing the regions in cluster 6 with respect to all our other regions.
# sending the data to the server
job_2 <- submitGreatJob(split_regions[["2"]], merged_peaks, species="hg38")
#Human GRCh38/hg38 is the genome we used
job_2
# we fetch the data back
res_2 <- getEnrichmentTables(job_2)
names(res_2)
```
Molecular function is about being enzymes, signalling proteins, biological process is what they are involved in and cellular component is where they work. This gives us an overview.

## Plotting the top Biological Processes, molecular functions and cellular components of Cluster 2

```{r, fig.width=9, fig.height=6}
# looking at the biological processes
bp <- res_2$`GO Biological Process`
head(bp)
# reorder it from most to least statistically significant before being alphabetically ordered.
ggplot(head(bp,20), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c()

# looking at the molecular function
mf <- res_2$`GO Molecular Function`
head(mf)
ggplot(head(mf,20), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c()

# looking at cellular components
cc <- res_2$`GO Cellular Component`
ggplot(head(cc,20), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c()
```
We get different genotology terms and it is telling us how many regions in our background are associated with this term and how many regions in our foreground, so the genes of interest are associated with this.
Fold Enrichment is how much more we have in our region of interest than in the background.

On the x-axis is the strength of enrichment and the color of the dot represents its significance.
Smaller sets need more enrichment in order to be significant than large sets. That's why we look at both the significance and the fold enrichment which indicates the magnitude of enrichment.
Disclaimer: not everything makes sense because the same genes get reused in many biological processes, but we can learn something about what the clusters do.

Here we analyzed cluster 2 which involves mainly CREB3, so looking at our plots we can get an idea of how CREB3 is involved in following processes in the cell. 
- Biological processes: We can see a relative high probability that cluster 2 is involved in response to stimulus, phosphorouse metabolic processes and positive regulation of cell differentiation. 
- Molecular function: Cluster 2 seems to be relatively highly involved in small molecule binding, identical protein binding, transcription factor activity and transcription factor binding as well as protein binding. 
- Cellular components: Cluster 2 can mainly be found in the cell, cell parts, cytoplasm, nuclear lumen and parts of the membrane. 


### Do another Cluster - Cluster 9

Now we do the same analysis for cluster 9.

```{r}
job_9 <- submitGreatJob(split_regions[["9"]], merged_peaks, species="hg38")
#Human GRCh38/hg38 is the genome we used
job_9
# we fetch the data back
res_9 <- getEnrichmentTables(job_9)
names(res_9)
```
## Plotting the top Biological Processes, molecular functions and cellular components of Cluster 9

```{r, fig.width=9, fig.height=6}
# looking at the biological processes
bp <- res_9$`GO Biological Process`
head(bp)
# reorder it from most to least statistically significant before being alphabetically ordered.
ggplot(head(bp,20), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c()

# looking at the molecular function
mf <- res_9$`GO Molecular Function`
head(mf)
ggplot(head(mf,20), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c()

# looking at cellular components
cc <- res_9$`GO Cellular Component`
ggplot(head(cc,20), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c()
```

Cluster 9 mainly involves CREB3 and CREB3L1 so looking at the plots we can see how these are involved in following functions.

- Biological processes: Cluster 9 seems to be highly involved in protein complex biogenesis, protein complex assebly as well as protein targeting and establishment of protein localization to organelle. 
- Molecular function: Looking at the Enrichement plot we can see that cluster 9 has molecular functions in active transmembrane transporter activity, cofactor binding, peptide binding as well as amide binding. 
- Cellular components: Cluster 9 can mainly be found in Organells, extracellular regions and parts of extracellular regions. 

### Do another Cluster - Cluster 7

Now we do the same analysis for cluster 7.

```{r}
job_7 <- submitGreatJob(split_regions[["7"]], merged_peaks, species="hg38")
#Human GRCh38/hg38 is the genome we used
job_7
# we fetch the data back
res_7 <- getEnrichmentTables(job_7)
names(res_7)
```

## Plotting the top Biological Processes, molecular functions and cellular components of Cluster 7

```{r, fig.width=9, fig.height=6}
# looking at the biological processes
bp <- res_7$`GO Biological Process`
head(bp)
# reorder it from most to least statistically significant before being alphabetically ordered.
ggplot(head(bp,20), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c()

# looking at the molecular function
mf <- res_7$`GO Molecular Function`
head(mf)
ggplot(head(mf,20), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c()

# looking at cellular components
cc <- res_7$`GO Cellular Component`
ggplot(head(cc,20), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c()
```

Cluster 7 mainly involves CREB3L and CREB1 as we figured out with our colored enriched Heatmaps so the following functions may be tightly correlated with CREB3L and CREB1. 

- Biological processes: Cluster 7 seems to play a role in cell cylce and glycosal compound metabolic process. 
- Molecular function: Protein homodimerization activity and protein dimerization activity may be regulated by cluster 7. 
- Cellular components: CREB3L1 and CREB1 are mainly found in organelles, nucleus, intracellular membrane bound organelles and into intracellular organelles. 


All the results above are to be enjoyed with caution and will be further evaluated and compared with existing knowlege about the different CREB family members in our results and interpretation section. 

```{r}
sessionInfo()
```
