---
title: "Project"
author: "Sarah Eiben, Michelle Baumgarnter, Deni Hodzic"
date: "`r Sys.Date()`"
output: html_document
---


# CREB family member and cofactor binding in K562 cells

Outline:
- Introduction
- Results
- Methods
- Discussion --> Still needs to be added
- References


## Introduction

CREB is an important element in the cAMP signaling pathway, which is bound to the DNA in its inactive state. CREB is then activated by active catalytic domains and ATP molecules. After conversion, phosphorylated CREB can bind to different cofactors, which stimulates the transcription of targeted genes. Although, we know that various CREB family members exist, only the function of CREB1 is understood in more detail, yet.

The K562 cell line is a human cell line derived from myelogenous leukemia cells. The fact that these cells are cancer cells and therefore immortal cells enables the researchers to analyze and use them for scientific purposes over multiple decades.

In our Group project we aim to find the different binding domains of CREB1, CREB3, CREB3L1 and CREB5. By comparing these which each other and in a further step looking at the interaction with H3K27ac and CREB co-factors (CREBBP, JUN, EP300, HDAC2) at these discovered domains, we eventually hope to understand the function of the individual CREB family members better.


# Install the newest version of epiwraps

```{r}
BiocManager::install("ETHZ-INS/epiwraps")
```

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(Rsubread)
  library(rtracklayer)
  library(Biostrings)
  library(Rfastp)
  library(epiwraps)
})
```

# Downloading the data
In this file all the data is being downloaded from ENCODE.
The data we use is CREB peaks and signal p-values (CREB1, CREB3, CREB3L1, CREB5), cofactor peaks and signal p-values (CREBBP, JUN, EP300, HDAC2) and acetylation peaks and signal p-values (H3K27ac). We saved them into different folders.

## Downloading CREB peaks
These are all IDR thresholded bed narrowpeaks, saved in a folder named peaks_creb.

```{r}
dir.create("peaks_creb")
# CREB1
download.file("https://www.encodeproject.org/files/ENCFF193LLN/@@download/ENCFF193LLN.bed.gz", dest="peaks_creb/peaks_creb1.bed.gz")
# CREB3
download.file("https://www.encodeproject.org/files/ENCFF829IEE/@@download/ENCFF829IEE.bed.gz", dest="peaks_creb/peaks_creb3.bed.gz")
# CREB3L1
download.file("https://www.encodeproject.org/files/ENCFF118GMS/@@download/ENCFF118GMS.bed.gz", dest="peaks_creb/peaks_creb3l1.bed.gz")
# CREB5
download.file("https://www.encodeproject.org/files/ENCFF875JMR/@@download/ENCFF875JMR.bed.gz", dest="peaks_creb/peaks_creb5.bed.gz")
```

## Downloading CREB signal p-value data
These are bigwig files saved in a folder named signal_pv_creb.

```{r}
dir.create("signal_pv_creb")
# CREB1
download.file("https://www.encodeproject.org/files/ENCFF014CIM/@@download/ENCFF014CIM.bigWig", dest=("signal_pv_creb/signal_pv_creb1.bw"), mode="wb")
# CREB3
download.file("https://www.encodeproject.org/files/ENCFF462EMG/@@download/ENCFF462EMG.bigWig", dest=("signal_pv_creb/signal_pv_creb3.bw"), mode="wb")
# CREB3L1
download.file("https://www.encodeproject.org/files/ENCFF638ZCD/@@download/ENCFF638ZCD.bigWig", dest=("signal_pv_creb/signal_pv_creb3l1.bw"), mode="wb")
# CREB5
download.file("https://www.encodeproject.org/files/ENCFF335JJU/@@download/ENCFF335JJU.bigWig", dest=("signal_pv_creb/Signal_pv_creb5.bw"), mode="wb")
```

## Downloading cofactor peaks
Downloading the peaks of the following cofactors: CREBBP, JUN, EP300 and HDAC2 into a folder named peaks_cofactors.
Maybe we won't need these - then we will just delete this chunk again.

```{r}
dir.create("peaks_cofactors")
# CREBBP
download.file("https://www.encodeproject.org/files/ENCFF532VPN/@@download/ENCFF532VPN.bed.gz", dest="peaks_cofactors/peaks_crebbp.bed.gz")
# JUN
download.file("https://www.encodeproject.org/files/ENCFF865UPM/@@download/ENCFF865UPM.bed.gz", dest="peaks_cofactors/peaks_jun.bed.gz")
# EP300
download.file("https://www.encodeproject.org/files/ENCFF702XPO/@@download/ENCFF702XPO.bed.gz", dest="peaks_cofactors/peaks_ep300.bed.gz")
# HDAC2
download.file("https://www.encodeproject.org/files/ENCFF150FJT/@@download/ENCFF150FJT.bed.gz", dest="peaks_cofactors//peaks_hdac2.bed.gz")
```

## Downloading cofactor signal p-value
Download of cofactor signal p-values (CREBBP, JUN, EP300 and HDAC2) in bigwig format into a folder named signal_pc_covactors.

```{r}
dir.create("signal_pv_cofactors")
# CREBBP
download.file("https://www.encodeproject.org/files/ENCFF783COJ/@@download/ENCFF783COJ.bigWig", dest="signal_pv_cofactors/signal_pv_crebbp.bw", mode="wb")
# JUN
download.file("https://www.encodeproject.org/files/ENCFF220RDW/@@download/ENCFF220RDW.bigWig", dest="signal_pv_cofactors/signal_pv_jun.bw", mode="wb")
# EP300
download.file("https://www.encodeproject.org/files/ENCFF325DSL/@@download/ENCFF325DSL.bigWig", dest="signal_pv_cofactors/signal_pv_ep300.bw", mode="wb")
# HDAC2
download.file("https://www.encodeproject.org/files/ENCFF240UFL/@@download/ENCFF240UFL.bigWig", dest="signal_pv_cofactors/signal_pv_hdac2.bw", mode="wb")
```

## Downloading H3K27ac peaks and signal p-value
Downloading the Acetylation peaks (bed narrow peaks) and p-values (bigwig) of H3K27ac into a folder named H2K27ac.

```{r}
dir.create("H3K27ac")
# H3K27ac peaks
download.file("https://www.encodeproject.org/files/ENCFF437DPT/@@download/ENCFF437DPT.bed.gz", dest="H3K27ac/h3k27ac_peaks.bed.gz")
# H3K27ac signal p-value
download.file("https://www.encodeproject.org/files/ENCFF469JMR/@@download/ENCFF469JMR.bigWig", dest="H3K27ac/h3k27ac_signal_pv.bw", mode="wb")
```
# Reading in the peak data
In this chunk we are reading in the peak data in order to use it later.

```{r}
# CREB peaks
peaks_creb1 <- import("peaks_creb/peaks_creb1.bed.gz", format="NarrowPeak")
peaks_creb3 <- import("peaks_creb/peaks_creb3.bed.gz", format="NarrowPeak")
peaks_creb3l1 <- import("peaks_creb/peaks_creb3l1.bed.gz", format="NarrowPeak")
peaks_creb5 <- import("peaks_creb/peaks_creb5.bed.gz", format="NarrowPeak")
# Cofactor peaks
peaks_crebbp <- import("peaks_cofactors/peaks_crebbp.bed.gz", format="NarrowPeak")
peaks_jun <- import("peaks_cofactors/peaks_jun.bed.gz", format="NarrowPeak")
peaks_ep300 <- import("peaks_cofactors/peaks_ep300.bed.gz", format="NarrowPeak")
peaks_hdac2 <- import("peaks_cofactors/peaks_hdac2.bed.gz", format="NarrowPeak")
# H3K27ac peaks
peaks_h3k27ac <- import("H3K27ac/h3k27ac_peaks.bed.gz", format="NarrowPeak")
```


# Reading in the signal p-value data

This step is added in order to follow our train of thoughts and process.
First we downloaded and read everything in without the mode="bw" which does not work on windows. Hence we had to redo this time-consuming step and add the mode.

```{r}
# try something pierre told me to:
download.file("https://www.encodeproject.org/files/ENCFF783COJ/@@download/ENCFF783COJ.bigWig", dest="signal_pv_cofactors/signal_pv_crebbp.bw", mode="wb")
# CREB signal p-value
pv_creb1 <- import.bw("signal_pv_cofactors/signal_pv_crebbp.bw", format="bw")
```


Eventually, all the data necessary for our group project was easily obtainable using the ENCODE-project.


```{r}
suppressPackageStartupMessages({
  library(GenomicRanges)
  library(epiwraps)
  library(ggplot2)
  library(rGREAT)
  library(AnnotationHub)
  library(Rsubread)
  library(rtracklayer)
  library(Biostrings)
  library(Rfastp)
  library(ensembldb)
  library(bsseq)
  library(BiocParallel)
  library(edgeR)
  library(DMRcate)
})
```

# Region access

Using the functions regionOverlaps and regionUpset we analyse the access of the DNA by the different CREB family members.

First we do everything just for Chromosome 1 to get an idea, then we move on to the whole genome (project part 6).

## Restricting all of the CREB family memeber peaks to just chromosome 1

```{r}
# CREB1
peaks_creb1_chr1 <- peaks_creb1[seqnames(peaks_creb1)=="chr1"]
# CREB3
peaks_creb3_chr1 <- peaks_creb3[seqnames(peaks_creb3)=="chr1"]
# CREB3L1
peaks_creb3l1_chr1 <- peaks_creb3l1[seqnames(peaks_creb3l1)=="chr1"]
# CREB5
peaks_creb5_chr1 <- peaks_creb5[seqnames(peaks_creb5)=="chr1"]
```

## Getting an idea of the peaks in chromosome 1

Should we do this??

Using just the strongest peaks.

```{r}
#look at how many peaks there are
length(peaks_creb1_chr1)
head(peaks_creb1_chr1)
#My peaks for some reason doesn't have the qValue?
strong_peaks_creb1 <- peaks_creb1_chr1[peaks_creb1_chr1$score>median(peaks_creb1_chr1$score)]
#extend shows you a little more area around the peak, the resolution is 20, specify the regions you want to analyse
peakSignals_creb1_chr1 <- signal2Matrix("signal_pv_creb/signal_pv_creb1", regions=strong_peaks_creb1, extend=2000, w=20)
class(peakSignals_creb1_chr1[[1]])
#the data itself is a big matrix, let's look at it
peakSignals_creb1_chr1[[1]][1:5, 1:5]
plotEnrichedHeatmaps(peakSignals_creb1_chr1, raster_by_magick=FALSE)
```

## Creating a regionOverlap and a regionUpset for all CREB family members for chromosome 1

First we create an Overlap of the peaks of chromosome 1 of all CREB family members and an Upsetplot of the same regions.

Look at lecture 5 code for this!

```{r}
listofRegions <- list(creb1=peaks_creb1_chr1, creb3=peaks_creb3_chr1, creb3l1=peaks_creb3l1_chr1, creb5=peaks_creb5_chr1)
regionOverlaps(listofRegions)
regionUpset(listofRegions)
```

## Restricting to CREB1 sites

Here we tried out how the upsetplot changes when we restrict it to CREB1.

```{r}
regionUpset(listofRegions, reference = peaks_creb1_chr1)
```

## Restricting to CREB3 sites

Here we tried out how the upsetplot changes when we restrict it to CREB3.

```{r}
regionUpset(listofRegions, reference = peaks_creb3_chr1)
```

## Restricting to CREB3L1 sites

Here we tried out how the upsetplot changes when we restrict it to CREB3L1.

```{r}
regionUpset(listofRegions, reference = peaks_creb3l1_chr1)
```
## Restricting to CREB5 sites
Here we tried out how the upsetplot changes when we restrict it to CREB5.

```{r}
regionUpset(listofRegions, reference = peaks_creb5_chr1)
```

# How do these plots make sense?

We can see in the RegionOverlap plot that CREB3L1 and CREB1 have the most overlapping regions followed by CREB3L1 and CREB3, then CREB1 and CREB3. CREB5 does not seem to overlap much with any of the other regions.

In the unrestricted RegionUpset these findings are reproduced and you can see additionally that CREB1, CREB3L1 and CREB3 have many common intersections that are shared amongst the 3 of them.
CREB1 and CREB3L1 share by far the most intersections.

When restricting the RegionUpset plot to CREB1 you can see that it shares most intersections with CREB3L1 followed by CREB3. The intersection number of CREB3L1, CREB1 and CREB3 is again shown to be quite high.

Restriction to CREB3 shows that intersections with CREB3L1 are highest, followed by CREB1 and then CREB5.

The order intersections of CREB3L1 is from high to low: CREB1, CREB3, CREB5, where the intersections with CREB5 are almost inexistent.

CREB5 seems to have very few intersections with the other family members but most with CREB3.

Intersection size is generally decreasing in the following order: CREB3L1, CREB1, CREB3, CREB5.


# Clustering of the signals and enrichment analysis

Our second approach to compare the different family members with each other is to cluster the signals and do an enrichment analysis with the clusters. After plotting the signals in a combined heatmap, the signals will be clustered and analyzed in an enrichment analysis.

Help for this in the code from lecture 9.

```{r}
suppressPackageStartupMessages({
  library(GenomicRanges)
  library(epiwraps)
  library(ggplot2)
  library(rGREAT)
  library(AnnotationHub)
  library(Rsubread)
  library(rtracklayer)
  library(Biostrings)
  library(Rfastp)
  library(ensembldb)
  library(bsseq)
  library(BiocParallel)
  library(edgeR)
  library(DMRcate)
})
```
## Plotting the signals

As regions we are going to be using the peaks from the different CREB family members that we downloaded in Project_2 and the tracks are the corresponding signal p-values.
We make a list of the signal p-values of the CREB family members.
Then we also make a list of the peak values and merge them together to create one long list of non-overlapping regions.
Finally we combine these in a matrix and create a heatmap.

```{r}
# create a list with all the peaks on chromosome 1
peak_list <- GRangesList("Peaks CREB1"=peaks_creb1_chr1, "Peaks CREB3"=peaks_creb3_chr1, "Peaks CREB3L1"=peaks_creb3l1_chr1, "Peaks CREB5"=peaks_creb5_chr1, compress=TRUE)
# merge the peaks together to create one list of non-overlapping regions
merged_peaks <- reduce(unlist(peak_list))
# make a list of the signal p-values
pv_list <- list(CREB1="signal_pv_creb/signal_pv_creb1.bw", CREB3="signal_pv_creb/signal_pv_creb3.bw", CREB3L1="signal_pv_creb/signal_pv_creb3l1.bw", CREB5="signal_pv_creb/Signal_pv_creb5.bw")
# we obtain the matrix of the signal around the regions:
m_creb1 <- signal2Matrix(pv_list, merged_peaks)
# maybe set an extend and a width?
plotEnrichedHeatmaps(m_creb1)
```
Each row represents the same region across the heatmaps.

Here you can see that there are strongest peaks in CREB3L1, followed by similar looking CREB1 and CREB3 and rather weak looking CREB5.

Analyze this more (not sure if I can speak of strengths of peaks here or what would be correct?).

## Clustering

Now we use clustering to set rows that are similar together and seperate them from the rest.
Let's start by trying with 5 clusters. Then we change the number of clusters until we reach a plateau of the variance explained by clusters.


```{r}
cl <- clusterSignalMatrices(m_creb1, k=10)
# with 5 clusters 73% of the variance is explained by clusters
# with 4 clusters 64 % of the variance is explained by clusters
# with 6 clusters 77%
# with 7 clusters 80%
# with 8 clusters 82%
# with 9 clusters 84%
# with 10 clusters 85%
# with 11 clusters 86%
# so let's go with 10 clusters for a start and see what we can tell.
table(cl)
head(cl)
length(cl)
# The numbers at [1] tell you where your regions are: first region is in cluster 10 with 303 elements
merged_peaks$cluster <- cl
```

### Plotting the clusters

We can split the rows by clusters to visualize this.
The meaning of the cluster is dependent on the order of the matrix, which depends on the order of the rows in the regions.
It's really easy if one generates another regions list that is ordered  and one keeps the same cluster identities everything will be messed up!
We have to keep in mind that the orders of both of them need to be matched when something is changed.
Typically when clusters are exhausting most of the pattern in the data, the results are more stable. For reproducibility of workflows it is important to set the seed, we chose to set the seed at 123.

```{r}
set.seed(123)
plotEnrichedHeatmaps(m_creb1, row_split=cl)
```
Interpretation:

Cluster 1: Represented in all CREB family members except for CREB5.
Cluster 2: Weak representation in all but CREB3.
Cluster 3: A cluster of few regions strong in CREB1 and CREB3L1.
Cluster 4: Medium representation in CREB1, CREB3 and CREB3L1 and barely any representation in CREB5.
Cluster 5: Strong in CREB3 and one of the few that is also represented quite strongly in CREB5.
Cluster 6: Strong in CREB1 and CREB3L1.
Cluster 7: Similar to Cluster 6.
Cluster 8: Seems to be something especially strongly represented by CREB1.
Cluster 9: A cluster of few regions remarkably strong and broad peaked in CREB3.
Cluster 10: Similar to clusters 6 and 7 but broader and stronger peaks.

The curves at the top is the average signal across all the regions but with clusters we have multiple lines. But we cannot tell which one is which.
So we set colors.

#### Adding colors

```{r}
# create a vector of colors and have as names of the vector the names of our clusters (in this case just numbers).
mycolors <- c("1"="yellow", "2"="orange", "3"="red", "4"="pink", "5"="purple", "6"="blue", "7"="green", "8"="brown", "9"="gray", "10"="black")
plotEnrichedHeatmaps(m_creb1, row_split=cl, mean_color=mycolors, trim=c(0.95))
```
Bring the curves into the representation!

## Plotting just the averages

We tried to extract the average signals to get a more broad overview of the signals, we can also do this by cluster.
This gives us a data frame that has for each cluster for each sample or signal at each position the value of the mean signal.
We can use this to create a plot that tells us in each cluster what the average signals of each factor around the positions around the regions of interest are.

```{r}
d <- meltSignals(m_creb1, splitBy=cl)
ggplot(d, aes(position, mean, colour=sample)) + geom_line(size=1.2) + facet_wrap(~split)
```

These plots are exactly like the plots above the heat maps with the difference that instead of plotting the different lines for one signal together we are plotting the lines for one cluster but different signals together. You can plot whatever you want.

Clustering is dangerous because you might change around factors until you like the result.

From this plot we can generally tell once more that CREB5 is not very strongly involved in any of these peaks. All the other family members have some clusters that they are the strongest representative which means they might be strongly involved in whatever these processes represent.

More analyis here?

# Enrichment analysis

Now we want to figure out what the regions are. For example what are the regions in cluster 1, are they associated with or close to genes that have a different kind of function than the regions in cluster 5?
Typically you add more signals in order to find more patterns of correlation. Are they bound by other factors?

Find what's enriched in one cluster with respect to the others:

```{r}
# we first split the regions by cluster:
split_regions <- split(merged_peaks, cl)
# Important: when a job is submitted to GREAT a set of regions, a GRanges object, needs to be submitted BUT there also needs to be some sort of background added to the picture. The elements in a list of regions are always compared to other elements. So we are comparing the regions in cluster 6 with respect to all our other regions.
# sending the data to the server
job <- submitGreatJob(split_regions[["6"]], merged_peaks, species="hg38")
#Human GRCh38/hg38 is the genome we used
job
# we fetch the data back
res <- getEnrichmentTables(job)
names(res)
```
Molecular function is about being enzymes, signalling proteins, biological process is what they are involved in and cellular component is where they work. This gives us an overview.

## Plotting the top Biological Processes, molecular functions and cellular components

```{r, fig.width=9, fig.height=6}
# looking at the biological processes
bp <- res$`GO Biological Process`
head(bp)
# reorder it from most to least statistically significant before being alphabetically ordered.
ggplot(head(bp,20), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c()
# looking at the molecular function
mf <- res$`GO Molecular Function`
head(mf)
ggplot(head(mf,20), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c()
# looking at cellular components
cc <- res$`GO Cellular Component`
ggplot(head(cc,20), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c()
```
We get different genotology terms and it is telling us how many regions in our background are associated with this term and how many regions in our foreground, so the genes of interest are associated with this.
Fold Enrichment is how much more we have in our region of interest than in the background.

On the x-axis is the strength of enrichment and the color of the dot represents its significance.
Smaller sets need more enrichment in order to be significant than large sets. That's why we look at both the significance and the fold enrichment which indicates the magnitude of enrichment.
Disclaimer: not everything makes sense because the same genes get reused in many biological processes, but we can learn something about what the clusters do.

Analyze this and link it to the biological background:
Biological processes:
Molecular function:
Cellular components:

Now we only did this for cluster 6 compared to all clusters. We need to do more comparisons like this and figure out what cluster stands for what.


# Adding the cofactors and one acetylation to the analysis

```{r}
# get the regions on the cofactors and the methylation for just chromosome 1
# CREBBP
peaks_crebbp_chr1 <- peaks_crebbp[seqnames(peaks_crebbp)=="chr1"]
# CREB3
peaks_jun_chr1 <- peaks_jun[seqnames(peaks_jun)=="chr1"]
# CREB3L1
peaks_ep300_chr1 <- peaks_ep300[seqnames(peaks_ep300)=="chr1"]
# CREB5
peaks_hdac2_chr1 <- peaks_hdac2[seqnames(peaks_hdac2)=="chr1"]
# H3K27ac
peaks_h3k27ac_chr1 <- peaks_h3k27ac[seqnames(peaks_h3k27ac)=="chr1"]
```

## Creating a regionOverlap and a regionUpset for all CREB family members for chromosome 1 AND the cofactors and acetylation

We make a list of regions that includes the cofactors and the acetylation to it and use this list to do another regionOverlaps and regionUpset plot.

```{r}
listofRegions_cf <- list(creb1=peaks_creb1_chr1, creb3=peaks_creb3_chr1, creb3l1=peaks_creb3l1_chr1, creb5=peaks_creb5_chr1, crebbp=peaks_crebbp, jun=peaks_jun, ep300=peaks_ep300, hdac2=peaks_hdac2, h3k27ac=peaks_h3k27ac)
regionOverlaps(listofRegions_cf)
regionUpset(listofRegions_cf)
```
What stands out in the RegionOverlaps plot is that the cofactors overlap strongly with one another and the overlaps of the cofactors with the CREB family members is in around the same dimension except for CREB5, which shows lower overlaps.
CREBBP seems to have not very many overlapping regions with anything.

The RegionUpset plot shows that the acetylations have by far the biggest intersection size and most intersections are also shown with its involvement. This acetylation seems to be very present.

### Creating different regionOverlap plots and regionUpset plots to see it more clearly

Following we are creating more specific plots by focusing on one cofactor or acetylation. We add our focus into the list of CREB family members and leave the others away.
To analyze these we can set the reference to different CREB family members which might give us some different information.

```{r}
# CREBBP
listofRegions_crebbp <- list(creb1=peaks_creb1_chr1, creb3=peaks_creb3_chr1, creb3l1=peaks_creb3l1_chr1, creb5=peaks_creb5_chr1, crebbp=peaks_crebbp_chr1)
regionOverlaps(listofRegions_crebbp)
regionUpset(listofRegions_crebbp, reference = peaks_crebbp_chr1)
```
CREBBP seems not to have very many overlaps with any of the CREB family members on chromosome 1.

```{r}
# JUN
listofRegions_jun <- list(creb1=peaks_creb1_chr1, creb3=peaks_creb3_chr1, creb3l1=peaks_creb3l1_chr1, creb5=peaks_creb5_chr1, jun=peaks_jun_chr1)
regionOverlaps(listofRegions_jun)
regionUpset(listofRegions_jun, reference = peaks_jun_chr1)
```
JUN has most overlaps with CREB1 and CREB3L1.


```{r}
# EP300
listofRegions_ep300 <- list(creb1=peaks_creb1_chr1, creb3=peaks_creb3_chr1, creb3l1=peaks_creb3l1_chr1, creb5=peaks_creb5_chr1, ep300=peaks_ep300_chr1)
regionOverlaps(listofRegions_ep300)
regionUpset(listofRegions_ep300, reference = peaks_ep300_chr1)
```
EP300 seems to have intersections with all CREB family members, most with CREB1, then CREB3L1, then CREB3 and even with CREB5.

```{r}
# HDAC2
listofRegions_hdac2 <- list(creb1=peaks_creb1_chr1, creb3=peaks_creb3_chr1, creb3l1=peaks_creb3l1_chr1, creb5=peaks_creb5_chr1, hdac2=peaks_hdac2_chr1)
regionOverlaps(listofRegions_hdac2)
regionUpset(listofRegions_hdac2, reference = peaks_hdac2_chr1)
```
HDAC2 has a similar overlap pattern to EP300.

```{r}
# H3K27ac
listofRegions_h3k27ac <- list(creb1=peaks_creb1_chr1, creb3=peaks_creb3_chr1, creb3l1=peaks_creb3l1_chr1, creb5=peaks_creb5_chr1, h3k27ac=peaks_h3k27ac_chr1)
regionOverlaps(listofRegions_h3k27ac)
regionUpset(listofRegions_h3k27ac, reference = peaks_h3k27ac_chr1)
regionUpset(listofRegions_h3k27ac, reference = peaks_creb1_chr1)
# comparing H3K27ac to the cofactors
listofRegions_h3k27ac_2 <- list(crebbp=peaks_crebbp_chr1, jun=peaks_jun_chr1, ep300=peaks_ep300_chr1, hdac2=peaks_hdac2_chr1, h3k27ac=peaks_h3k27ac_chr1)
regionOverlaps(listofRegions_h3k27ac_2)
regionUpset(listofRegions_h3k27ac_2, reference = peaks_h3k27ac_chr1)
```

H3K27ac seems to intersect most with CREB1, then CREB3L1, CREB3 and barely with CREB5. From the cofactors it has noticeable intersections with hdac2 and ep300.


## Doing analysis according to clusters

Let's look at how the different clusters interact with the cofactors and the methylation.

```{r}
# create a list with all the peaks on chromosome 1
peak_list_cf <- GRangesList("Peaks CREB1"=peaks_creb1_chr1, "Peaks CREB3"=peaks_creb3_chr1, "Peaks CREB3L1"=peaks_creb3l1_chr1, "Peaks CREB5"=peaks_creb5_chr1, "Peaks CREBBP"=peaks_crebbp_chr1, "Peaks JUN"=peaks_jun_chr1, "Peaks EP300"=peaks_ep300_chr1, "Peaks HDAC2"=peaks_hdac2_chr1, "Peaks H3K27ac"=peaks_h3k27ac_chr1, compress=TRUE)
# merge the peaks together to create one list of non-overlapping regions
merged_peaks_cf <- reduce(unlist(peak_list_cf))
# make a list of the signal p-values
pv_list_cf <- list(CREB1="signal_pv_creb/signal_pv_creb1.bw", CREB3="signal_pv_creb/signal_pv_creb3.bw", CREB3L1="signal_pv_creb/signal_pv_creb3l1.bw", CREB5="signal_pv_creb/Signal_pv_creb5.bw", CREBBP="signal_pv_cofactors/signal_pv_crebbp.bw", JUN="signal_pv_cofactors/signal_pv_jun.bw", EP300="signal_pv_cofactors/signal_pv_ep300.bw", HDAC2="signal_pv_cofactors/signal_pv_hdac2.bw", H3K27ac="H3K27ac/h3k27ac_signal_pv.bw")
# we obtain the matrix of the signal around the regions:
m_creb1_cf <- signal2Matrix(pv_list_cf, merged_peaks_cf)
# maybe set an extend and a width?
plotEnrichedHeatmaps(m_creb1_cf, row_title="Clustered analysis")
```

As you can see in the average plots, the basic normalization based on library size wasn't quite sufficient. While the size of the peak in the middle depends on the amount of the protein, we would expect the background around it to flatten to a similar level across experiments, which isn't the case.

Thus we try to correct for this by trying two approaches.

### Approach 1: Plot groups of experiments and combine

The first approach is to plot two groups of experiments seperately, so that they have their own color scale and then join them afterwarts.

```{r}
plotEnrichedHeatmaps(m_creb1_cf[1:6], scale_title="density\nCREB members", colors=c("white","darkred"), row_title="Approach 1") + plotEnrichedHeatmaps(m_creb1_cf[7:9], scale_title="density\nothers", row_title="Approach 1")
```

### Approach 2: Re-normalization

The second approach is to re-normalize the signals. This can be done with the function renormalizeBorders (2a).

```{r}
# Approach 2a
plotEnrichedHeatmaps(renormalizeBorders(m_creb1_cf), row_title = "Approach 2")
```

Approach 1 shows more clearly what we want to see, so we continue with that approach.

### Cluster

```{r}
cl_2 <- clusterSignalMatrices(m_creb1_cf, k=5)
# 10 clusters explain 96% of the variance
# 5 clusters explain 91% of the variance
merged_peaks_cf$cluster <- cl_2
```
```{r}
set.seed(124)
plotEnrichedHeatmaps(m_creb1_cf[1:6], row_split = cl_2, scale_title="density\nCREB members", colors=c("white","darkred"), row_title="Approach 1 with clusters") + plotEnrichedHeatmaps(m_creb1_cf[7:9], row_split = cl_2, scale_title="density\nothers", row_title="Approach 1 with clusters")
```

Interpret this.


# Doing the previous things for the whole genome

We are trying to do this for the whole genome:

```{r}
listofRegions_whole <- list(creb1=peaks_creb1, creb3=peaks_creb3, creb3l1=peaks_creb3l1, creb5=peaks_creb5)
regionOverlaps(listofRegions_whole)
regionUpset(listofRegions_whole)
```
## Comparison to chromosome 1

In order to be able to compare it better to what only chromosome 1 looked like, we are adding the chromosome 1 UpsetPlot here as well.

```{r}
listofRegions <- list(creb1=peaks_creb1_chr1, creb3=peaks_creb3_chr1, creb3l1=peaks_creb3l1_chr1, creb5=peaks_creb5_chr1)
regionOverlaps(listofRegions)
regionUpset(listofRegions)
```
Compare them:

## Restricting to CREB1 sites

Here we tried out how the upsetplot changes when we restrict it to CREB1.

```{r}
regionUpset(listofRegions_whole, reference = peaks_creb1)
```
## Restricting to CREB3 sites

Here we tried out how the upsetplot changes when we restrict it to CREB3.

```{r}
regionUpset(listofRegions_whole, reference = peaks_creb3)
```

## Restricting to CREB3L1 sites

Here we tried out how the upsetplot changes when we restrict it to CREB3L1.

```{r}
regionUpset(listofRegions_whole, reference = peaks_creb3l1)
```

## Restricting to CREB3L1 sites

Here we tried out how the upsetplot changes when we restrict it to CREB3L1.

```{r}
regionUpset(listofRegions_whole, reference = peaks_creb3l1)
```

## Restricting to CREB5 sites
Here we tried out how the upsetplot changes when we restrict it to CREB5.

```{r}
regionUpset(listofRegions_whole, reference = peaks_creb5)
```
# How do these plots make sense?

Looking at the plots above we can find the most interactions between CREB1 and CREB3l1, followed by CREB3 and CREB3l1, the least interactions are found with CREB5, which mainly interacts with CREB3. 

# Clustering of the signals and enrichment analysis

## Plotting the signals

```{r}
# create a list with all the peaks on the whole genome
peak_list_whole <- GRangesList("Peaks CREB1"=peaks_creb1, "Peaks CREB3"=peaks_creb3, "Peaks CREB3L1"=peaks_creb3l1, "Peaks CREB5"=peaks_creb5, compress=TRUE)
# merge the peaks together to create one list of non-overlapping regions
merged_peaks_whole <- reduce(unlist(peak_list))
# make a list of the signal p-values
pv_list_whole <- list(CREB1="signal_pv_creb/signal_pv_creb1.bw", CREB3="signal_pv_creb/signal_pv_creb3.bw", CREB3L1="signal_pv_creb/signal_pv_creb3l1.bw", CREB5="signal_pv_creb/Signal_pv_creb5.bw")
# we obtain the matrix of the signal around the regions:
m_creb1_whole <- signal2Matrix(pv_list_whole, merged_peaks_whole)
# maybe set an extend and a width?
plotEnrichedHeatmaps(m_creb1_whole)
```

Here comes up the question how you want to analyze the whole genome. Going off of the whole genome for the heatmaps altogether makes not so much sense, since then you cannot divide it up into logical groups or assign it back to the original genome. So it would make more sense to cluster gene after gene and then compare them to eachother.

This was initially our goal but as we had neither enough processing power or time to let this all run for the whole genome so we selected one more chromosome based on this paper: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC387762/
The chosen chromosome would be chromosome 22, since according to this paper CREB1 binds to many sites on this chromosome. Let's find out how it is for the other CREB family members.

## Restricting all of the CREB family memeber peaks to chromosome 22

```{r}
# CREB1
peaks_creb1_chr22 <- peaks_creb1[seqnames(peaks_creb1)=="chr22"]
# CREB3
peaks_creb3_chr22 <- peaks_creb3[seqnames(peaks_creb3)=="chr22"]
# CREB3L1
peaks_creb3l1_chr22 <- peaks_creb3l1[seqnames(peaks_creb3l1)=="chr22"]
# CREB5
peaks_creb5_chr22 <- peaks_creb5[seqnames(peaks_creb5)=="chr22"]
```

## Creating a regionOverlap and a regionUpset for all CREB family members for chromosome 22

First we create an Overlap of the peaks of chromosome 22 of all CREB family members and an Upsetplot of the same regions.

```{r}
listofRegions_22 <- list(creb1=peaks_creb1_chr22, creb3=peaks_creb3_chr22, creb3l1=peaks_creb3l1_chr22, creb5=peaks_creb5_chr22)
regionOverlaps(listofRegions_22)
regionUpset(listofRegions_22)
```

# Restricting to CREB1

```{r}
regionUpset(listofRegions_22, reference = peaks_creb1_chr22)
```

## Restricting to CREB3 sites

```{r}
regionUpset(listofRegions_22, reference = peaks_creb3_chr22)
```

## Restricting to CREB3L1 sites

```{r}
regionUpset(listofRegions_22, reference = peaks_creb3l1_chr22)
```

## Restricting to CREB5 sites

```{r}
regionUpset(listofRegions_22, reference = peaks_creb5_chr22)
```
## Explain how these plots make sense? How do they compare to chr1?

# Adding the cofactors and one acetylation to the analysis

```{r}
# get the regions on the cofactors and the methylation for just chromosome 22
# CREBBP
peaks_crebbp_chr22 <- peaks_crebbp[seqnames(peaks_crebbp)=="chr22"]
# CREB3
peaks_jun_chr22 <- peaks_jun[seqnames(peaks_jun)=="chr22"]
# CREB3L1
peaks_ep300_chr22 <- peaks_ep300[seqnames(peaks_ep300)=="chr22"]
# CREB5
peaks_hdac2_chr22 <- peaks_hdac2[seqnames(peaks_hdac2)=="chr22"]
# H3K27ac
peaks_h3k27ac_chr22 <- peaks_h3k27ac[seqnames(peaks_h3k27ac)=="chr22"]
```

## Creating a regionOverlap and a regionUpset for all CREB family members for chromosome 22 AND the cofactors and acetylation

We make a list of regions that includes the cofactors and the acetylation to it and use this list to do another regionOverlaps and regionUpset plot.

```{r}
listofRegions_cf_22 <- list(creb1=peaks_creb1_chr22, creb3=peaks_creb3_chr22, creb3l1=peaks_creb3l1_chr22, creb5=peaks_creb5_chr22, crebbp=peaks_crebbp, jun=peaks_jun, ep300=peaks_ep300, hdac2=peaks_hdac2, h3k27ac=peaks_h3k27ac)
regionOverlaps(listofRegions_cf_22)
regionUpset(listofRegions_cf_22)
```
What stands out in this RegionOverlaps plot is that the cofactor hdac2 overlaps strongly with the acetylation h3k27ac. 

CREB5 seems to have the least overlaps.

The RegionUpset plot shows that the acetylations have by far the biggest intersection size and most intersections are also shown with its involvement. This acetylation seems to be very present.

### Creating different regionOverlap plots and regionUpset plots to see it more clearly

Following we are creating more specific plots by focusing on one cofactor or acetylation. We add our focus into the list of CREB family members and leave the others away.
To analyze these we can set the reference to different CREB family members which might give us some different information.

```{r}
# CREBBP
listofRegions_crebbp_22 <- list(creb1=peaks_creb1_chr22, creb3=peaks_creb3_chr22, creb3l1=peaks_creb3l1_chr22, creb5=peaks_creb5_chr22, crebbp=peaks_crebbp_chr22)
regionOverlaps(listofRegions_crebbp_22)
regionUpset(listofRegions_crebbp_22, reference = peaks_crebbp_chr22)
```

CREBBP seems not to have very many overlaps with any of the CREB family members on chromosome 22.

```{r}
# JUN
listofRegions_jun_22 <- list(creb1=peaks_creb1_chr22, creb3=peaks_creb3_chr22, creb3l1=peaks_creb3l1_chr22, creb5=peaks_creb5_chr22, jun=peaks_jun_chr22)
regionOverlaps(listofRegions_jun_22)
regionUpset(listofRegions_jun_22, reference = peaks_jun_chr22)
```
JUN seems not to have very many overlaps with any of the CREB family members on chromosome 22.

```{r}
# EP300
listofRegions_ep300_22 <- list(creb1=peaks_creb1_chr22, creb3=peaks_creb3_chr22, creb3l1=peaks_creb3l1_chr22, creb5=peaks_creb5_chr22, ep300=peaks_ep300_chr22)
regionOverlaps(listofRegions_ep300_22)
regionUpset(listofRegions_ep300_22, reference = peaks_ep300_chr22)
```
EP300 also seems to have little intersections with the CREB family members but most with CREB1. 

```{r}
# HDAC2
listofRegions_hdac2_22 <- list(creb1=peaks_creb1_chr22, creb3=peaks_creb3_chr22, creb3l1=peaks_creb3l1_chr22, creb5=peaks_creb5_chr22, hdac2=peaks_hdac2_chr22)
regionOverlaps(listofRegions_hdac2_22)
regionUpset(listofRegions_hdac2_22, reference = peaks_hdac2_chr22)
```

HDAC2 seems to have most intersections with the CREB family members especially CREB1 followed by CREB3l1. 

```{r}
# H3K27ac
listofRegions_h3k27ac_22 <- list(creb1=peaks_creb1_chr22, creb3=peaks_creb3_chr22, creb3l1=peaks_creb3l1_chr22, creb5=peaks_creb5_chr22, h3k27ac=peaks_h3k27ac_chr22)
regionOverlaps(listofRegions_h3k27ac_22)
regionUpset(listofRegions_h3k27ac_22, reference = peaks_h3k27ac_chr22)
regionUpset(listofRegions_h3k27ac_22, reference = peaks_creb1_chr22)
```
H3K27ac seems to intersect most with CREB3l1 and CREB1 then CREB3, and barely with CREB5.


## Doing analysis according to clusters

Let's look at how the different clusters interact with the cofactors and the methylation.

--> STILL NEEDS TO BE ADJUSTED NOT WORKING YET

```{r}
# create a list with all the peaks on chromosome 22
peak_list_cf_22 <- GRangesList("Peaks CREB1"=peaks_creb1_chr22, "Peaks CREB3"=peaks_creb3_chr22, "Peaks CREB3L1"=peaks_creb3l1_chr22, "Peaks CREB5"=peaks_creb5_chr22, "Peaks CREBBP"=peaks_crebbp_chr22, "Peaks JUN"=peaks_jun_chr22, "Peaks EP300"=peaks_ep300_chr22, "Peaks HDAC2"=peaks_hdac2_chr22, "Peaks H3K27ac"=peaks_h3k27ac_chr22, compress=TRUE)
# merge the peaks together to create one list of non-overlapping regions
merged_peaks_cf_22 <- reduce(unlist(peak_list_cf_22))
# make a list of the signal p-values
pv_list_cf <- list(CREB1="signal_pv_creb/signal_pv_creb1.bw", CREB3="signal_pv_creb/signal_pv_creb3.bw", CREB3L1="signal_pv_creb/signal_pv_creb3l1.bw", CREB5="signal_pv_creb/Signal_pv_creb5.bw", CREBBP="signal_pv_cofactors/signal_pv_crebbp.bw", JUN="signal_pv_cofactors/signal_pv_jun.bw", EP300="signal_pv_cofactors/signal_pv_ep300.bw", HDAC2="signal_pv_cofactors/signal_pv_hdac2.bw", H3K27ac="H3K27ac/h3k27ac_signal_pv.bw")
# we obtain the matrix of the signal around the regions:
m_creb1_cf_22 <- signal2Matrix(pv_list_cf, merged_peaks_cf_22)
# maybe set an extend and a width?
plotEnrichedHeatmaps(m_creb1_cf_22, row_title="Clustered analysis")
```
-->> ADD IN ANALAYSIS OF THE HEATMAPS, WHAT DO WE SEE HERE? 

### Approach 1: Plot groups of experiments and combine

The first approach is to plot two groups of experiments seperately, so that they have their own color scale and then join them afterwarts.

```{r}
plotEnrichedHeatmaps(m_creb1_cf_22[1:6], scale_title="density\nCREB members", colors=c("white","darkred"), row_title="Approach 1") + plotEnrichedHeatmaps(m_creb1_cf_22[7:9], scale_title="density\nothers", row_title="Approach 1")
```

### Approach 2: Re-normalization

The second approach is to re-normalize the signals. This can be done with the function renormalizeBorders (2a).

```{r}
# Approach 2a
plotEnrichedHeatmaps(renormalizeBorders(m_creb1_cf_22), row_title = "Approach 2")
```

Approach 1 shows more clearly what we want to see, so we continue with that approach.

### Cluster

```{r}
cl_2_22 <- clusterSignalMatrices(m_creb1_cf_22, k=5)
# 10 clusters explain 96% of the variance
# 5 clusters explain 91% of the variance
merged_peaks_cf_22$cluster <- cl_2_22
```
```{r}
set.seed(124)
plotEnrichedHeatmaps(m_creb1_cf_22[1:6], row_split = cl_2_22, scale_title="density\nCREB members", colors=c("white","darkred"), row_title="Approach 1 with clusters") + plotEnrichedHeatmaps(m_creb1_cf_22[7:9], row_split = cl_2_22, scale_title="density\nothers", row_title="Approach 1 with clusters")
```

Interpret this.
-->> COULD'T RUN THE ABOVE CHUNKS AS I COULDNT CREATE m_creb1_cf_22



# Methods

As previously mentioned, we used data provided from ENCODE from the K562 cells, as this data was already provided for all the CREB family members and cofactors that we needed.

Furthermore, we used most importantly the epiwraps package in Biocmanager and other packages and functions that were introduced to us in the lecture.

Inflicted through our rather slow computers and limited processing abilities, we were forced to cut back on some of our ideas.

Following is how we approached to find answers to our questions.
After downloading and importing the data from ENCODE into R, we proceeded to analyze the CREB family member peaks for chromosome 1 with the help of varying regionOverlaps and regionUpset plots. In a next step, we plotted heatmaps of the chromosome 1 signals. Clustering helped get a better overview and through enrichment analysis we tried to connect peak signals to functions.

Then we added the peak signals of cofactors, to get an idea of how their peaks coincide with the CREB family member peaks. Again regionOverlaps and regionUpset plots were used in several variations with different foci to figure out how the factors interplay.

Adding one specific acetylation to the analysis, we proceeded to look at how this interacts with the CREB family members and the cofactors.

The acetylation and the cofactor interactions were then also analyzed using clusters to see what factors could potentially be connected to eachother.

The plan was to then proceed with repeating all that we had done for the whole genome. This was partially done, but in our opinion it would result in most information to seperate the chromosomes and do the whole process for each chromosome to be able to assign the functions and results to corresponding places in the genome. As we realized this rather late in the process, we restricted ourselves to repeat the analysis for chromosome 22, as we found in literature that CREB1 highly interacts with this chromosome. We tried to figure out if other CREB family members, cofactors and the acetylation equally interact with chromosome 22 and then compare it to what we saw in chromosome 1.

This process was chosen according to our (limited) knowledge of the field and our knowledge was also used to analyze and compare the results.

# References

Following we list the references we used.

- Euskirchen, Ghia et al. “CREB binds to multiple loci on human chromosome 22.” Molecular and cellular biology vol. 24,9 (2004): 3804-14. doi:10.1128/MCB.24.9.3804-3814.2004
- https://en.wikipedia.org/wiki/CREB
- Sassone-Corsi, Paolo. “The cyclic AMP pathway.” Cold Spring Harbor perspectives in biology vol. 4,12 a011148. 1 Dec. 2012, doi:10.1101/cshperspect.a011148
- Mayr, B., Montminy , M. Transcriptional regulation by the phosphorylation-dependent factor CREB. Nat Rev Mol Cell Biol 2, 599–609 (2001). https://doi.org/10.1038/35085068
- Shaukat, Ammad et al. “Interplay Between BALL and CREB Binding Protein Maintains H3K27 Acetylation on Active Genes in Drosophila.” Frontiers in cell and developmental biology vol. 9 740866. 28 Sep. 2021, doi:10.3389/fcell.2021.740866
- https://www.encodeproject.org

```{r}
sessionInfo()

```

